### ISSUE TITLE
USB support for IMX8MP (dwc3)

### ISSUE THREAD

**Issue by hcl-dr** [barebox/barebox#13]
As of 2022.08 this is not working. Host driver panics on NXP IMX8MP Eval kit:
```
barebox@NXP i.MX8MP DR:/ usb
usb: USB: scanning bus for devices...
usb1: Bus 001 Device 001: ID 0000:0000 XHCI Host Controller
ERROR: xHCI xHCI0: XHCI timeout on event type 33... cannot recover.
BUG: failure at drivers/usb/host/xhci-ring.c:471/xhci_wait_for_event()!
BUG!
Call trace:
[<ffd802ec>] (unwind_backtrace+0x0/0x88) from [<ffd01d34>] (panic+0x5c/0x78)
[<ffd01d34>] (panic+0x5c/0x78) from [<ffd30264>] (xhci_wait_for_event+0xb8/0x11c)
[<ffd30264>] (xhci_wait_for_event+0xb8/0x11c) from [<ffd2e8ac>] (xhci_submit_control_msg+0x554/0xab0)
[<ffd2e8ac>] (xhci_submit_control_msg+0x554/0xab0) from [<ffd25740>] (usb_control_msg+0xb0/0xec)
[<ffd25740>] (usb_control_msg+0xb0/0xec) from [<ffd25fd4>] (usb_new_device+0xd8/0x6b0)
[<ffd25fd4>] (usb_new_device+0xd8/0x6b0) from [<ffd26ed0>] (usb_hub_detect+0x314/0x438)
[<ffd26ed0>] (usb_hub_detect+0x314/0x438) from [<ffd26630>] (usb_host_detect+0x84/0x88)
[<ffd26630>] (usb_host_detect+0x84/0x88) from [<ffd266d8>] (usb_rescan+0x60/0x6c)
[<ffd266d8>] (usb_rescan+0x60/0x6c) from [<ffd57708>] (do_usb+0x48/0x128)
[<ffd57708>] (do_usb+0x48/0x128) from [<ffd07178>] (execute_command+0x44/0x8c)
[<ffd07178>] (execute_command+0x44/0x8c) from [<ffd10418>] (run_list_real+0x8d4/0x980)
[<ffd10418>] (run_list_real+0x8d4/0x980) from [<ffd0f998>] (parse_stream_outer+0x144/0x1f0)
[<ffd0f998>] (parse_stream_outer+0x144/0x1f0) from [<ffd107f8>] (run_shell+0x64/0xac)
[<ffd107f8>] (run_shell+0x64/0xac) from [<ffd0186c>] (run_init+0x228/0x250)
[<ffd0186c>] (run_init+0x228/0x250) from [<ffd018e4>] (start_barebox+0x50/0x8c)
[<ffd018e4>] (start_barebox+0x50/0x8c) from [<ffd7fbb8>] (barebox_non_pbl_start+0x1a4/0x1ac)
[<ffd7fbb8>] (barebox_non_pbl_start+0x1a4/0x1ac) from [<ffd0000c>] (__bare_init_start+0x0/0x4)
[<ffd0000c>] (__bare_init_start+0x0/0x4) from [<40203b00>] (0x40203b00)
[<40203b00>] (0x40203b00) from [<40203740>] (0x40203740)
```

**Comment by a3f** [comment]
Did this work before?

**Comment by hcl-dr** [comment]
Not to my knowledge. Works on IMX8MQ, but there are some changes at the phy layer on the MP.
Since the MP eval board sports 6GB memory I also checked the Barebox layout which is confined within 4GB physical memory. The kernel code has a quirk for 64 bit DMA, but not sure if this of significance here.

**Comment by hcl-dr** [comment]
```
From a236673824ee7e3ddcb0acb0a055eddd5c607937 Mon Sep 17 00:00:00 2001
From: Hans Christian Lonstad <hcl@datarespons.com>
Date: Mon, 22 Aug 2022 08:34:18 +0200
Subject: [PATCH] USB DW3: Apply freerunning phy clock quirk

Fixes #13 (imx8mp)
---
 drivers/usb/dwc3/core.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index 30aaef90ac..7be3b18df9 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -1001,6 +1001,8 @@ static void dwc3_get_properties(struct dwc3 *dwc)
 	dwc->dr_mode = of_usb_get_dr_mode(dev->device_node, NULL);
 	dwc->hsphy_mode = of_usb_get_phy_mode(dev->device_node, NULL);
 
+	dwc->dis_u2_freeclk_exists_quirk = of_property_read_bool(dev->device_node,
+		"snps,dis-u2-freeclk-exists-quirk");
 	dwc->lpm_nyet_threshold = lpm_nyet_threshold;
 	dwc->tx_de_emphasis = tx_de_emphasis;
 
-- 
2.34.1
```

**Comment by hcl-dr** [comment]
Patched on v2022.08.0

**Comment by hcl-dr** [comment]
Host mode working (OTG is not)

**Comment by saschahauer** [comment]
Hi,
Could you send that patch to the mailing list or create a pull request from it?

Thanks
  Sascha

**Comment by hcl-dr** [comment]
Prepare pull request

**Comment by hcl-dr** [comment]
Fixed in v2022.10.0
